### PR TITLE
fix: hydrate store after mount

### PR DESCRIPTION
**Pull request by iamacook**
## What it solves

Hydration render errors when loading persisted state from `localStorage`.

## How this PR fixes it

Instead of initialising Redux with a `preloadedStore`, a `storeHydrator` HOC now wraps the app, hydrating the store after mount.

## How to test it

Open the address book and ensure entries exist. Refresh the page and observe no hydration errors.

### CODE REVIEW

File: src/components/address-book/AddressBookTable/index.tsx [hunk 1]
```diff
@@ -14,7 +14,7 @@ import TokenTransferModal from '@/components/tx/modals/TokenTransferModal'
 import css from './styles.module.css'
 import EthHashInfo from '@/components/common/EthHashInfo'
 import Box from '@mui/material/Box'
-import useAddressBookEntries from './useAddressBookEntries'
```
Comment (katspaugh) [inline src/components/address-book/AddressBookTable/index.tsx:17]: You should probably remove the file then?

File: src/pages/_app.tsx [hunk 2]
```diff
@@ -90,8 +89,8 @@ const SafeWebCore = ({ Component, pageProps }: AppProps): ReactElement | null =>
 
         <Notifications />
       </AppProviders>
-    </Provider>
+    </>
   )
 }
 
-export default SafeWebCore
+export default storeHydrator(SafeWebCore)
```
Comment (katspaugh) [inline src/pages/_app.tsx:96]: Any specific reason it needs to wrap the entire app as opposed to just calling it from InitApp or in AppProviders?

File: src/store/index.ts [hunk 3]
```diff
@@ -50,13 +52,31 @@ const persistedSlices: (keyof PreloadedState<RootState>)[] = [
...
+const makeStore = () => {
+  const hydrationReducer: typeof rootReducer = (state, action) => {
+    if (action.type === HYDRATE) {
+      return {
+        ...state,
+        ...action.payload,
+      }
+    }
+    return rootReducer(state, action)
+  }
```
Comment (katspaugh) [inline src/store/index.ts:68]: I would define this fn outside of makeStore.

File: src/store/storeHydrator.tsx [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import { Provider } from 'react-redux'
+import type { Store } from 'redux'
+
+import { getPersistedState, RootState } from '@/store'
+
+export const HYDRATE = '@@HYDRATE'
+
+export const createStoreHydrator = (makeStore: () => Store<RootState>) => {
+  return (Component: any) => {
+    return class HydrationWrapper extends React.Component<any, any> {
```
Comment (katspaugh) [inline src/store/storeHydrator.tsx:11]: Why does it need to be a class?

**Review by katspaugh — COMMENTED** [review]
Ultimate (and a very naive) question 😄 : why is this better than just creating a custom hook that reads from LS, and dispatches it to the entire store?

File: src/pages/_app.tsx (hunk 2 above)
Comment (iamacook) [reply src/pages/_app.tsx:96]: It has to call the hydration action before any other actions are dispatched.

File: src/store/storeHydrator.tsx (hunk 4 above)
Comment (iamacook) [reply src/store/storeHydrator.tsx:11]: I was encountering issues with the hydration action not dispatching first. I only got it working in a class component.

**Comment by iamacook** [comment]
> Ultimate (and a very naive) question 😄 : why is this better than just creating a custom hook that reads from LS, and dispatches it to the entire store?

As with my two other comments, the hydration action must be called before any others are dispatched. When using a custom hook, it was not consistently being dispatched first.

**Comment by usame-algan** [comment]
Two things I noticed from briefly testing it:

1. Cookie Banner pops up again on each reload
2. Dark mode checkbox is not checked even if `darkMode` is set to true

Could this be related to defering the `preloadedState`?

**Comment by iamacook** [comment]
> Two things I noticed from briefly testing it:
> 
>     1. Cookie Banner pops up again on each reload
> 
>     2. Dark mode checkbox is not checked even if `darkMode` is set to true
> 
> 
> Could this be related to defering the `preloadedState`?

1. I've added a close banner condition. The state was being hydrated but there was no close banner reaction present in the effect.
2. The initial value of dark mode in the store is `undefined`. I've set it to true. You will have to change the value in the settings so as to set it but then it will persist correctly.

File: src/hooks/__tests__/useChainId.test.ts [hunk 5]
```diff
@@ -68,9 +68,10 @@ describe('useChainId hook', () => {
       query: {},
     }))
 
-    renderHook(() => useAppDispatch()(setLastChainId('100')))
-
-    const { result } = renderHook(() => useChainId())
+    const { result } = renderHook(() => {
+      useAppDispatch()(setLastChainId('100'))
+      return useChainId()
+    })
```
Comment (iamacook) [inline src/hooks/__tests__/useChainId.test.ts:74]: I don't understand how this was passing before. Every call to `render` or `renderHook` now creates a new store, but they were previously resetting the store.

Any insights on this @mikhailxyz?

File: src/components/common/CookieBanner/index.tsx [hunk 6]
```diff
@@ -101,6 +101,8 @@ const CookieBanner = (): ReactElement | null => {
   useEffect(() => {
     if (shouldOpen) {
       dispatch(openCookieBanner({}))
+    } else {
+      dispatch(closeCookieBanner())
```
Comment (iamacook) [inline src/components/common/CookieBanner/index.tsx:105]: The cookie banner previously remained open as we didn't handle this case.

File: src/store/settingsSlice.ts [hunk 7]
```diff
@@ -24,7 +24,7 @@ const initialState: SettingsState = {
     qr: true,
   },
   theme: {
-    darkMode: undefined,
+    darkMode: true,
```
Comment (iamacook) [inline src/store/settingsSlice.ts:27]: Dark mode will remain unchecked until you toggle it. The persisted slice will currently be `undefined` in your `localStorage`.

**Review by katspaugh — APPROVED** [review]
Lovely jobly, that's not a trivial thing!

**Comment by katspaugh** [comment]
@iamacook have you seen this btw?

https://github.com/vercel/next.js/blob/canary/examples/with-redux-persist/pages/_app.js

It looks simple on the first glance. There's a `blacklist` config for slices too.

**Comment by iamacook** [comment]
> @iamacook have you seen this btw?
> 
> https://github.com/vercel/next.js/blob/canary/examples/with-redux-persist/pages/_app.js
> 
> It looks simple at first glance. There's a `blacklist` config for slices too.

Yes, I had great difficulty with it. It stores the entire store under one key. In order to use it, it needs `next-redux-wrapper` as well.

File: src/hooks/__tests__/useLoadChains.test.ts [hunk 8]
```diff
@@ -21,10 +21,9 @@ jest.mock('@gnosis.pm/safe-react-gateway-sdk', () => {
   }
 })
 
-// Jest tests for the useInitChains hook
-describe('useInitChains hook', () => {
+// Jest tests for the useLoadChains hook
+describe('useLoadChains hook', () => {
   beforeEach(() => {
-    jest.resetModules()
```
Comment (iamacook) [inline src/hooks/__tests__/useLoadChains.test.ts:27]: `resetModules` resets all modules. It breaks the test and is not needed.

**Review by usame-algan — APPROVED** [review]
Nice work on this 💯 🚀

**Pull request merged**